xclim/core/bootstrapping.py

```diff
@@ -0,0 +1,236 @@
+        we may want to calculate the percentiles of each day for a base periode, for example the first 30 years,
+        then we use theses percentiles to see how, in the whole periode, the daily exceedance rate evolves.
+        The exceedance rate of a year is the number of days where they individually exceed the thershold (the percentile) for this day of the year.
+        For example, if the 95 percentile for the 5 march is 10 °C it means 95% of 5 march in the base period have an averaged tempareture below 10°C.
+        Once this is estimated for each day of a year (eventually skipping the 29 feb), we can use theses values to calculate exceedane rate of eac year.
+        This can give for example, that the year 2021 has 50 days exceeding the 95 percentile.
+        Plotting the evolution of this count for each years gives a good view of how the extreme values are becoming more or less frequent.
+
+        However this analysis introduces a bias because the percentile is computed using the data of the base period and is also used for the calculation of exceedance rates for the same base period.
+        Thus the years of the base period are affected by sampling variability.
+
+        To workaround this issue in the in-base period, we can apply the following algorithm
+        The data set of in-base period will thereafter be called ds
+        1. Split ds in an in-base of 29 years and a single year of out of base period
+        2. Add a virtual year in the constructed in-base period by repeating one year.
+        3. Get the thershold (your percentile of interest) form this 30 years in-base period
+        4. Calculate the exceedance rate of the out base period (the single year)
+        5. Repeat 28 times the steps 2, 3, 4 but each time, the repeated year of step 2 must be another year of the in-base
+        6. To obtain the final time serie of exceedance rate for the out of base year, average the 29 estimates obtained
+
+        7. Repeat the whole process for each year of the base period
+
+        TODO: show how to bootstrap a custom indice
+
+    """
+    in_base_period = da.sel(time=config.in_base_slice)
+    in_base_exceedance_rates = _bootstrap_period(
+        in_base_period, config, exceedance_function
+    )
+    if config.out_of_base_slice is None:
+        return in_base_exceedance_rates
+    out_of_base_period = da.sel(time=config.out_of_base_slice)
+    in_base_threshold = _calculate_thresholds(in_base_period, config)
+    out_of_base_exceedance = _calculate_exceedances(
+        config, exceedance_function, out_of_base_period, in_base_threshold
+    )
+    return xr.concat([in_base_exceedance_rates, out_of_base_exceedance], dim="time")
+
+
+def _bootstrap_period(
+    ds_in_base_period: DataArray,
+    config: BootstrapConfig,
+    exceedance_function: ExceedanceFunction,
+) -> DataArray:
+    period_exceedance_rates = []
+    for year_ds in ds_in_base_period.groupby("time.year"):
```

Comment:
We can probably parallelize this with dask, but I'm not sure how. 
We can't simply chunk the ds by time and let it run, it won't change the performances because each iteration needs the whole ds in base period to perform the computation of percentile.
Maybe by calling dask delay instead of appending result in the list ?

Reply:
AFAIU, this would be already parallelized, no? I mean, assuming each call to `_bootstrap_year` returns a lazy DataArray and then `out` is the concatenation of them. When computation is triggered, dask will construct it's tree and _then_ parallelize it as much as it can.

On another note, I think I would use:
```suggestion
    for year in np.unique(ds_in_base_period.time.dt.year):
```
I have a mistrust of `groupby` when it comes to dask performance. Here you are only using the year values, so no need to invoke the whole groupby machinery!

Reply:
I'm very new to Dask but I think it will parallelize between the different chunks if the DataArray is chunked, but will not parallelize this for loop because we treat multiple time the whole DA, no ?
I tried something like :
```
for year_ds in ds_in_base_period.groupby("time.year"):
        period_exceedance_rates.append(
            dask.delayed(_bootstrap_year)(
                ds_in_base_period.copy(deep=True),
                year_ds[0],
                config,
                exceedance_function,
            )
        )
    out = xr.concat(dask.compute(*period_exceedance_rates), dim="time")
```
It seems to work fine on small DA but goes into deadlock with bigger ones. I still have to investigate the exact reason why.
I will also have a look at map_blocks, maybe it can help here.

Reply:
Resolving this because with the new implementation of @huard it is now outdated